lib/elixir_analyzer/test_suite/take_a_number_deluxe.ex

```diff
@@ -19,13 +20,37 @@ defmodule ElixirAnalyzer.TestSuite.TakeANumberDeluxe do
     end
   end
 
-  feature "uses @impl GenServer" do
+  check_source "uses @impl GenServer" do
     type :actionable
-    find :any
     comment Constants.take_a_number_deluxe_annotate_impl_genserver()
 
-    form do
-      @impl GenServer
+    check(%Source{code_ast: code_ast}) do
+      {_, %{defs_without_impls: defs_without_impls}} =
+        Macro.postwalk(code_ast, %{defs_without_impls: [], impl?: false}, &find_defs_and_impls/2)
+
+      defs_without_impls == []
     end
   end
+
+  @def_ops [:def, :defp, :defmacro, :defmacrop, :defguard, :defguardp]
+  @genserver_callbacks_in_this_exercise [:init, :handle_call, :handle_cast, :handle_info]
+  defp find_defs_and_impls(node, acc) do
+    acc =
+      case node do
+        {:@, _, [{:impl, _, [{:__aliases__, _, [:GenServer]}]}]} ->
+          %{acc | impl?: true}
+
+        {op, _, [{function_name, _, _} | _]} when op in @def_ops ->
+          if function_name in @genserver_callbacks_in_this_exercise and !acc.impl? do
+            %{acc | impl?: false, defs_without_impls: [function_name | acc.defs_without_impls]}
+          else
+            %{acc | impl?: false}
+          end
+
+        _ ->
+          acc
```

Comment:
Should we return `%{acc | impl?: false}` here instead?
Otherwise this case wouldn't get flagged:
```elixir
@impl GenServer
defp some_helper_function, do: :something
def handle_call(...)
```

Reply:
I mean, arguably that example might be fine, but there might be worse ones

Reply:
By returning `%{acc | impl?: false}` here, we're saying that `@impl` only counts if it's the last thing before `def`, right? That doesn't sound correct to me 🤔. If I compile this snippet:
```
defmodule FooBar do
  use GenServer

  @impl GenServer
  @doc "some doc"
  @spec this_is_not_a_genserver_callback :: nil
  def this_is_not_a_genserver_callback, do: nil
end
```

I still get a warning, which means that `@impl` counts:
```
warning: got "@impl GenServer" for function this_is_not_a_genserver_callback/0 but this behaviour does not specify such callback. The known callbacks are:
```

Reply:
OK, I stand corrected.